internal/hcsoci/cmd.go

```diff
@@ -0,0 +1,318 @@
+package hcsoci
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sys/windows"
+)
+
+// Cmd represents a command being prepared or run in a process host.
+type Cmd struct {
+	// Host is the process host in which to launch the process.
+	Host cow.ProcessHost
+
+	// OS is the operating system of the process host. FUTURE: Move to cow.ProcessHost.
+	OS string
+
+	// InUVM is true if Host is a utility VM. FUTURE: Move to cow.ProcessHost.
+	InUVM bool
+
+	// The OCI spec for the process.
+	Spec *specs.Process
+
+	// Standard IO streams to relay to/from the process.
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+
+	// Log provides a logrus entry to use in logging IO copying status.
+	Log *logrus.Entry
+
+	// Context provides a context that terminates the process when it is done.
+	Context context.Context
+
+	// CopyAfterExitTimeout is the amount of time after process exit we allow the
+	// stdout, stderr relays to continue before forcibly closing them if not
+	// already completed. This is primarily a safety step against the HCS when
+	// it fails to send a close on the stdout, stderr pipes when the process
+	// exits and blocks the relay wait groups forever.
+	CopyAfterExitTimeout time.Duration
+
+	// Process is filled out after Start() returns.
+	Process cow.Process
+
+	// ExitState is filled out after Wait() (or Run() or Output()) completes.
+	ExitState *ExitState
+
+	iogrp     errgroup.Group
+	stdinErr  atomic.Value
+	allDoneCh chan struct{}
+}
+
+// ExitState contains whether a process has exited and with which exit code.
+type ExitState struct {
+	exited bool
+	code   int
+}
+
+// ExitCode returns the exit code of the process, or -1 if the exit code is not known.
+func (s *ExitState) ExitCode() int {
+	if !s.exited {
+		return -1
+	}
+	return s.code
+}
+
+// ExitError is used when a process exits with a non-zero exit code.
+type ExitError struct {
+	*ExitState
+}
+
+func (err *ExitError) Error() string {
+	return fmt.Sprintf("process exited with exit code %d", err.ExitCode())
+}
+
+// Additional fields to hcsschema.ProcessParameters used by LCOW
+type lcowProcessParameters struct {
+	hcsschema.ProcessParameters
+	OCIProcess *specs.Process `json:"OciProcess,omitempty"`
+}
+
+// escapeArgs makes a Windows-style escaped command line from a set of arguments
+func escapeArgs(args []string) string {
+	escapedArgs := make([]string, len(args))
+	for i, a := range args {
+		escapedArgs[i] = windows.EscapeArg(a)
+	}
+	return strings.Join(escapedArgs, " ")
+}
+
+// Command makes a Cmd for a given command and arguments.
+func Command(host cow.ProcessHost, os string, name string, arg ...string) *Cmd {
```

Comment:
If `ProcessHost` had an `OS() string` method we would not needed to pass `os` here.

Reply:
I mean that's exactly what the FUTURE things say.

Reply:
Yes but I would of had to read that first...